### PR TITLE
Correctly handling reference cycles between messages

### PR DESCRIPTION
**Pull request by wk8**
Previously, a message referencing itself, such as for example:
```
message Foo {
    string name = 1;
    repeated Foo bar = 2;
}
```
or more generally messages depending on one another so as to form a dependency cycle
would result in an infinite loop when trying to generate JSON schemas
from them.

This patch fixes that by creating JSON schema references in those cases.

Added unit tests on the new behaviour.

### CODE REVIEW

**Comment by wk8** [comment]
@chrusty : apologies for the big diff: I had to upgrade `github.com/alecthomas/jsonschema` enough to have the `Extras` field in `Type` (https://github.com/alecthomas/jsonschema/blob/7152f22193c99d748515db8612354f96c5e2c71a/reflect.go#L76), which in turn required to turn `Type.Properties` into a `*orderedmap.OrderedMap` instead of a regular map, which in turn resulted in quite a lot of re-ordering in JSON maps in the test output. If you look at the diff in tests though, it's only that: re-ordering. There is no change in existing behaviours.

**Comment by wk8** [comment]
Fixes:
https://github.com/chrusty/protoc-gen-jsonschema/issues/20
https://github.com/chrusty/protoc-gen-jsonschema/issues/19
https://github.com/chrusty/protoc-gen-jsonschema/issues/7

**Comment by chrusty** [comment]
Thank you for this effort @wk8. I'll give it a try and hopefully merge it in soon.

File: internal/converter/types.go [hunk 1]
```diff
@@ -61,11 +62,10 @@ componentLoop:
 }
 
 // Convert a proto "field" (essentially a type-switch with some recursion):
-func (c *Converter) convertField(curPkg *ProtoPackage, desc *descriptor.FieldDescriptorProto, msg *descriptor.DescriptorProto) (*jsonschema.Type, error) {
-
+func (c *Converter) convertField(curPkg *ProtoPackage, desc *descriptor.FieldDescriptorProto, msg *descriptor.DescriptorProto, duplicatedMessages map[*descriptor.DescriptorProto]string) (*jsonschema.Type, error) {
 	// Prepare a new jsonschema.Type for our eventual return value:
 	jsonSchemaType := &jsonschema.Type{
-		Properties: make(map[string]*jsonschema.Type),
+		Properties: orderedmap.New(),
```
Comment (chrusty) [inline internal/converter/types.go:68]: If you were to remove this line then we would no longer have the empty `"properties": {},` cruft in the generated schemas.

How do you feel about that? I think it is a good idea to keep the generated schemas as clean as possible. You'd have to clean up the schemas in the unit tests though, which I know is a horrible job.

**Comment by wk8** [comment]
@chrusty : thanks for the quick review, amended as requested :)

**Comment by wk8** [comment]
(also, just a quick thought: updating unit tests could be made much simpler/painless by just having the tests actually run `protoc` to re-generate `jsonschemas/*` files, and then check that they were no changes? Not gonna do that in that patch, but just an idea :) )

**Comment by wk8** [comment]
Thank you for the quick review @chrusty !

**Pull request merged**